technic/machines/HV/quarry.lua

```diff
@@ -184,402 +71,485 @@ local function can_dig_node(pos, node_name, owner, digger)
 	end
-	set_quarry_status(pos)
+	if fields.mesecons then
+		meta:set_string("mesecons", fields.mesecons)
+	end
+	if fields.channel then
+		meta:set_string("channel", fields.channel)
+	end
+	if fields.enable then meta:set_int("enabled", 1) end
+	if fields.disable then meta:set_int("enabled", 0) end
+	if fields.restart then reset_quarry(meta) end
+	update_formspec(meta)
+end
+
+local function add_particle_line(pos1, pos2, player)
+	local dist = vector.distance(pos1, pos2)
+	minetest.add_particlespawner({
+		playername = player:get_player_name(),
+		amount = dist * 80,  -- About 4 particles per node at any given time
+		time = 10,
+		minpos = pos1,
+		maxpos = pos2,
+		minsize = 0.8,
+		maxsize = 0.8,
+		minexptime = 0.5,
+		maxexptime = 0.5,
+		texture = "technic_line_particle.png",
+		glow = 14,
+	})
 end
 
-local digiline_def = function(pos, _, channel, msg)
+local function show_working_area(pos, _, player)
+	if not player or player:get_wielded_item():get_name() ~= "" then
+		-- Only spawn particles when using an empty hand
+		return
+	end
+	local meta = minetest.get_meta(pos)
+	local radius = meta:get_int("size") + 0.5
+	local offset = vector.new(meta:get_int("offset_x"), 0, meta:get_int("offset_z"))
+	-- Four top corners of the area
+	local p1 = vector.add(pos, vector.new(offset.x + radius, -0.5, offset.z + radius))
+	local p2 = vector.add(pos, vector.new(offset.x - radius, -0.5, offset.z + radius))
+	local p3 = vector.add(pos, vector.new(offset.x - radius, -0.5, offset.z - radius))
+	local p4 = vector.add(pos, vector.new(offset.x + radius, -0.5, offset.z - radius))
+	-- Draw the top square
+	add_particle_line(p1, p2, player)
+	add_particle_line(p2, p3, player)
+	add_particle_line(p3, p4, player)
+	add_particle_line(p4, p1, player)
+	-- Draw the edges down to the bottom
+	local down = vector.new(0, -meta:get_int("max_depth"), 0)
+	add_particle_line(p1, vector.add(p1, down), player)
+	add_particle_line(p2, vector.add(p2, down), player)
+	add_particle_line(p3, vector.add(p3, down), player)
+	add_particle_line(p4, vector.add(p4, down), player)
+end
+
+local function digiline_action(pos, _, channel, msg)
 	local meta = minetest.get_meta(pos)
 	if channel ~= meta:get_string("channel") then
 		return
 	end
-	-- Convert string messages to tables:
+	-- Convert string message to table
 	if type(msg) == "string" then
-		local smsg = msg:lower()
-		msg = {}
-		if smsg == "get" then
-			msg.command = "get"
-		elseif smsg:sub(1,7) == "radius " then
-			msg.command = "radius"
-			msg.value = smsg:sub(8,-1)
-		elseif smsg:sub(1,10) == "max_depth " then
-			msg.command = "max_depth"
-			msg.value = smsg:sub(11,-1)
-		elseif smsg == "on" then
-			msg.command = "on"
-		elseif smsg == "off" then
-			msg.command = "off"
-		elseif smsg == "restart" then
-			msg.command = "restart"
+		msg = msg:lower()
+		if msg == "get" or msg == "on" or msg == "off" or msg == "restart" then
+			msg = {command = msg}
+		elseif msg:sub(1, 7) == "radius " then
```

Comment:
Could reduce digiline action time a bit by using bit better performing find instead of sub, LuaJIT does not optimize this case. Potentially reduces penalty a bit when mesecons_debug is enabled and many quarries connected to network.
I've not tested actual penalty but it is not only performance, find is also shorter and reduces potential index errors (you can see immediately if it is correct or not without counting characters or thinking about indexing method).

So faster, cleaner and reduces human error surface:
```suggestion
		elseif msg:find("^radius ") then
```

Reply:
Interesting, I would have assumed that getting part of a string would be faster than finding a pattern 🤔

Reply:
Did a quick test running code in-game (not sure how accurate it is), comparing the two methods for 1,000,000 calls:

- `string.sub` with LuaJIT on: 337μs
- `string.find` with LuaJIT on: 71795μs
- `string.sub` with LuaJIT off: 20788μs
- `string.find` with LuaJIT off: 70662μs

Unless I'm doing something wrong with testing, it seems like my assumption is correct 🤔

Reply:
Yes, you're almost sure doing it wrong and JIT is optimizing out test code for you removing most of it. You need to use dynamic input to prevent caching, input caching isn't possible in actual use cases.
To prevent optimizing out code you can add counter for input string and even better make every other case fail in test loop with alternating input data.

Reply:
Also tested with actual minetest engine and found out that performance was actually very different from plain luajit run and with mt string.sub clearly won, very interesting thing was that sub was nearly 10x faster with all mods enabled but only 2x with mods disabled.
Did quick search for overrides as that would immediately explain it but didn't find any.

What you got seems right with mt engine and sub indeed performs better and probably should be left as is (your JIT results still seems skewed which is probably optimization).

Reply:
Did another test with this code, trying to make the input dynamic a bit:
```lua
local t = minetest.get_us_time()
for i=1, 100000 do
    local s = string.rep(i % 2 == 1 and "a" or "b", (i % 10) + 10)

    local x = s:sub(1, 5)
    -- OR
    local x = s:find("^aaaaa")
end
print(minetest.get_us_time() - t)
```
Subtracting the time it takes to run the loop without either `sub` or `find`, the times are much closer, but still `sub` wins.

Either way, we're dealing with microseconds of difference when it comes to the actual use case, so it really doesn't matter which one is used.

Reply:
Yeah it seems sub is faster anyway and true that this really does not need optimization for performance by itself.

Actually started discussion only for mesecons_debug penalty mechanism because it is a lot more sensitive (nothing wrong with that) and small things can make difference there: even just few % faster already allows commanding more devices.

> Subtracting the time it takes to run the loop without either sub or find, the times are much closer

Remember that code which creates input string is same for both and increases execution time, for accurate performance measurement. After getting results you'd also have to subtract time used up by utility stuff or it will skew proportions.